src/mobu/services/github_ci/ci_manager.py

```diff
@@ -0,0 +1,337 @@
+
+        We'd rather have false-negative GitHub checks than forever-in-progress
+        checks, because:
+        * There is no way to re-run in-progress checks from the GitHub UI
+        * There is no way to know for sure from the GitHub UI that a check
+          will never be concluded.
+
+        A failed check can easily be re-run by any user, so this is an
+        acceptable tradeoff, assuming false-negatives happen infrequently.
+
+        Scenarios:
+
+        ### Job completes
+
+        1. Mobu is told to shut down
+        2. Mobu tells GitHub that job has failed due to restart
+        3. Job finishes before mobu is SIGKILLed
+        4. Mobu tells GitHub that job is successful
+        5. Mobu is SIGKILLed or exits cleanly
+
+        Result: GitHub check run displays success, which is corrcect.
+                An incorrect failure status will have been displayed for a
+                brief period of time.
+        User action needed: None
+
+        ### Job does not complete
+
+        1. Mobu is told to shut down
+        2. Mobu tells GitHub that job has failed due to restart
+        3. Mobu is SIGKILLed
+
+        Result: GitHub check run displays failure status, which is correct.
+        User action needed: Re-run check in GitHub UI
+
+        ### Job completes, but mobu is killed before it can tell GitHub
+
+        This situation should be pretty rare.
+        1. Mobu is told to shut down
+        2. Mobu tells GitHub that job has failed due to restart
+        3. Job finishes before mobu is SIGKILLed
+        4. Mobu is SIGKILLed before it can tell GitHub the job has succeeded
+
+        Result: GitHub check run displays failure status, which is incorrect.
+        User action needed: Re-run check in GitHub UI
+        """
+        self._logger.info("Stopping CI manager...")
+
+        # Tell workers with in-progress jobs to stop after their current job
+        for worker in self.workers:
+            worker.stop()
+
+        # Tell GitHub all checks for in progress jobs are failed in case we
+        # shutdown uncleanly
+        awaits = [
+            worker.current_job.check_run.fail(error=self.shutdown_error_msg)
+            for worker in self.workers
+            if worker.current_job is not None
+        ]
+        await asyncio.gather(*awaits)
+
+        # Tell GitHub all checks queued are failed
+        awaits = []
+        while not self._queue.empty():
+            item = await self._queue.get()
+            if item is not None:
+                awaits.append(
+                    item.job.check_run.fail(error=self.shutdown_error_msg)
+                )
+        await asyncio.gather(*awaits)
+        self.lifecycle.marked_remaining.set()
+
+        # Tell workers listening on a currently empty queue to stop
+        for _ in self.workers:
+            await self._queue.put(None)
+
+        # Wait for workers to finish any in-progress jobs
+        for job in self._jobs:
+            await job.wait()
```

Comment:
Future work here too: we should probably just shut down the notebook immediately rather than wait for the jobs to finish.